halo2_debug/src/display.rs

```diff
@@ -0,0 +1,351 @@
+use ff::PrimeField;
+use halo2_middleware::circuit::{ColumnMid, VarMid};
+use halo2_middleware::expression::{Expression, Variable};
+use halo2_middleware::{lookup, shuffle};
+use num_bigint::BigUint;
+use std::collections::HashMap;
+use std::fmt;
+
+/// Wrapper type over `PrimeField` that implements Display with nice output.
+/// - If the value is a power of two, format it as `2^k`
+/// - If the value is smaller than 2^16, format it in decimal
+/// - If the value is bigger than congruent -2^16, format it in decimal as the negative congruent
+/// (between -2^16 and 0).
+/// - Else format it in hex without leading zeros.
+pub struct FDisp<'a, F: PrimeField>(pub &'a F);
+
+impl<F: PrimeField> fmt::Display for FDisp<'_, F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let v = (*self.0).to_repr();
+        let v = v.as_ref();
+        let v = BigUint::from_bytes_le(v);
+        let v_bits = v.bits();
+        if v_bits >= 8 && v.count_ones() == 1 {
+            write!(f, "2^{}", v.trailing_zeros().unwrap_or_default())
+        } else if v_bits < 16 {
+            write!(f, "{}", v)
+        } else {
+            let v_neg = (F::ZERO - self.0).to_repr();
+            let v_neg = v_neg.as_ref();
+            let v_neg = BigUint::from_bytes_le(v_neg);
+            let v_neg_bits = v_neg.bits();
+            if v_neg_bits < 16 {
+                write!(f, "-{}", v_neg)
+            } else {
+                write!(f, "0x{:x}", v)
+            }
+        }
+    }
+}
+
+/// Wrapper type over `Expression` that implements Display with nice output.
```

Comment:
The reason for the wrapper is to be able to include auxiliary information that allows displaying the expressions with column names.
In the `Expression` type we don't hold the column names, just the column type and index.  Then in `ConstraintSystemMid` we have a `HashMap` for column names.
`ExprDisp` can have a reference to this `HashMap` so that we can show the columns by name, like in this example https://github.com/privacy-scaling-explorations/halo2/blob/458ad7090eaf28f91e2218297c0f026dc7a12cd8/halo2_proofs/tests/compress_selectors.rs#L465-L468

Without these extra data, we could only print generic names like `advice0, fixed2, instance3`.  I think being able to use named columns makes the expression much more readable.